_posts/2022-04-26-type-based-optimizations-in-the-jit.md

```diff
@@ -0,0 +1,809 @@
+of the instruction at `L38`. If not, control is transferred to the
+failure label `label_9` (label `9` in the BEAM code):
+
+```assembly
+    cmp rdi, rsi
+    short jl L38
+    jmp label_9
+```
+
+Next is the call to the generic routine for comparing two arbitrary
+Erlang terms and a conditional jump to the failure label `label_9`
+if the first term is greater than or equal to the second term:
+
+```assembly
+L37:
+    call 4415504480
+    jge label_9
+```
+
+Thus, without type information, 12 instructions are needed to implement
+`is_lt`. Only 5 of those instructions are responsible for comparing
+the integer operands.
+
+Now let's see what happens when types are available:
+
+```erlang
+lt(A, B) when is_integer(A), is_integer(B) ->
+    .
+    .
+    .
+```
+
+When compiled by the compiler in OTP 25, the BEAM code looks like this:
+
+```
+    {test,is_integer,{f,7},[{x,0}]}.
+    {test,is_integer,{f,7},[{x,1}]}.
+    {test,is_lt,{f,9},[{tr,{x,0},{t_integer,any}},{tr,{x,1},{t_integer,any}}]}.
+    {move,{atom,smaller},{x,0}}.
+    return.
+  {label,9}.
+    {move,{atom,larger_or_equal},{x,0}}.
+    return.
+
+```
+
+The operands for the `is_lt` instruction now have types. The BEAM
+registers `{x,0}` and `{x,1}` have the type `{t_integer,any}`, which
+means an integer with an unknown range.
+
+Having that knowledge of the types, the JIT can emit a slightly
+shorter test for a small integer:
+
+```assembly
+# simplified small test since all other types are boxed
+    mov eax, edi
+    and eax, esi
+    test al, 1
+    short je L39
+```
+
+To do a better job, the JIT will need better type information. For example:
+
+```erlang
+map_size_lt(Map1, Map2) ->
+    if
+        map_size(Map1) < map_size(Map2) -> smaller;
+        true -> larger_or_equal
+    end.
+```
+
+The BEAM code looks like this:
+
+```
+    {gc_bif,map_size,{f,12},2,[{x,0}],{x,0}}.
+    {gc_bif,map_size,{f,12},2,[{x,1}],{x,1}}.
+    {test,is_lt,
+          {f,12},
+          [{tr,{x,0},{t_integer,{0,288230376151711743}}},
+           {tr,{x,1},{t_integer,{0,288230376151711743}}}]}.
+    {move,{atom,smaller},{x,0}}.
+    return.
+  {label,12}.
+    {move,{atom,larger_or_equal},{x,0}}.
+    return.
+```
+
+Both operands for `is_lt` now have the type
+`{t_integer,{0,288230376151711743}}`, meaning an integer in the range
+0 through 288230376151711743 (that is, `(1 bsl 58) - 1`). There is no
+documented upper limit for the number of elements in a map, but for
+the foreseeable future, there is no way that the number of elements in
+a map will exceed or even get close to 288230376151711743.
+
+Since both the lower and upper bounds for `{x,0}` and `{x,1}` fit in
+60 bits, there is no need to test the type of the operands:
+
+
+```assembly
+# is_lt_fss
+    mov rsi, qword ptr [rbx+8]
+    mov rdi, qword ptr [rbx]
+# skipped test for small operands since they are always small
+    cmp rdi, rsi
+    short jl L41
+    jmp label_12
```

Comment:
As an aside, it looks like we can shave off 5-8 bytes here by inverting the condition.

Reply:
Yes, that should be safe to do in OTP 25.

We could also shave off one instruction in the "unoptimized" case:

```assembly
# is_lt_fss
    mov rsi, qword ptr [rbx+8]
    mov rdi, qword ptr [rbx]
    mov eax, edi
    and eax, esi
    and al, 15
    cmp al, 15
    short jne L37
    cmp rdi, rsi
    short jmp L38
L37:
    call 4415504480
L38:
    jge label_9
L38:
```

Reply:
I have done the suggested optimization in erlang/otp#5895 and updated the example in the blog post.